### PR TITLE
Feature/c feature extraction

### PR DESCRIPTION
**Pull request by JacobToftgaardRasmussen**

### CODE REVIEW

File: c-svc23Vectors/math.c [hunk 1]
```diff
@@ -1,4 +1,6 @@
 #include "math.h"
+#include "types.h"
+#include <stdio.h>
```
Comment (humleflue) [inline c-svc23Vectors/math.c:3]:
```suggestion
```

File: c-svc23Vectors/math.c [hunk 2]
```diff
@@ -34,3 +36,94 @@ int dotProductVectors(const int *input1, const int *input2, int length) {
 int dotProduct3dVectors(const int input1[3], const int input2[3]) {
     return dotProductVectors((const int*)input1, (const int*)input2, 3);
 }
+
+double getBiggestNumber(double axisArray[], int length){
+    double highestNumber = 0;
```
Comment (humleflue) [inline c-svc23Vectors/math.c:41]:
```suggestion
    int highestNumber = 0;
```

File: c-svc23Vectors/math.c [hunk 3]
```diff
@@ -34,3 +36,94 @@ int dotProductVectors(const int *input1, const int *input2, int length) {
 int dotProduct3dVectors(const int input1[3], const int input2[3]) {
     return dotProductVectors((const int*)input1, (const int*)input2, 3);
 }
+
+double getBiggestNumber(double axisArray[], int length){
```
Comment (humleflue) [inline c-svc23Vectors/math.c:40]:
```suggestion
int getBiggestAbsNumber(int axisArray[], int length){
```

File: c-svc23Vectors/math.c [hunk 4]
```diff
@@ -34,3 +36,94 @@ int dotProductVectors(const int *input1, const int *input2, int length) {
...
+    double highestNumber = 0;
+    int negativeOrPositive = 1;
```
Comment (humleflue) [inline c-svc23Vectors/math.c:42]:
```suggestion
```

File: c-svc23Vectors/math.c [hunk 5]
```diff
@@ -34,3 +36,94 @@ int dotProductVectors(const int *input1, const int *input2, int length) {
...
+    int i;
+    for (i = 0; i < TIME_SERIES_VECTORS_AMOUNT; i++)
+    {
+        double currentNumber = axisArray[i];
```
Comment (humleflue) [inline c-svc23Vectors/math.c:46]:
```suggestion
        int currentNumber = abs(axisArray[i]);
```
Implementer en abs funktion.

File: c-svc23Vectors/math.c [hunk 6]
```diff
@@ -34,3 +36,94 @@ int dotProductVectors(const int *input1, const int *input2, int length) {
...
+            
+        } else
+        {
+            if (-1 * currentNumber > highestNumber)
+            {
+                highestNumber = -1 * currentNumber;
+                negativeOrPositive = -1;
+            }
+            
+        }
```
Comment (humleflue) [inline c-svc23Vectors/math.c:63]:
```suggestion
    if (currentNumber > highestNumber) {
        highestNumber = currentNumber;
    }
```

File: c-svc23Vectors/math.c [hunk 7]
```diff
@@ -34,3 +36,94 @@ int dotProductVectors(const int *input1, const int *input2, int length) {
...
+    }
+    return negativeOrPositive * highestNumber;
+}
+
+double getSmallestNumber(double axisArray[], int length){
```
Comment (humleflue) [inline c-svc23Vectors/math.c:68]: Abstract over biggest- and smallestNumber, as they are almost the same implementation. Maybe you can make an auxiliary function?

File: c-svc23Vectors/math.c [hunk 8]
```diff
@@ -34,3 +36,94 @@ int dotProductVectors(const int *input1, const int *input2, int length) {
...
+                smallestNumber = -1 * currentNumber;
+                negativeOrPositive = -1;
+            }
+            
+        }
+    }
+    return negativeOrPositive * smallestNumber;
+}
+
+double absMean(double axisArray[], int length){
```
Comment (humleflue) [inline c-svc23Vectors/math.c:96]:
```suggestion
int absMean(int axisArray[], int length){
```

File: c-svc23Vectors/math.c [hunk 9]
```diff
@@ -34,3 +36,94 @@ int dotProductVectors(const int *input1, const int *input2, int length) {
...
+    double result = absSum(axisArray, length);
```
Comment (humleflue) [inline c-svc23Vectors/math.c:97]:
```suggestion
    int result = absSum(axisArray, length);
```

File: c-svc23Vectors/math.c [hunk 10]
```diff
@@ -34,3 +36,94 @@ int dotProductVectors(const int *input1, const int *input2, int length) {
...
+    return result / length;    
+}
+
+double mean(double axisArray[], int length){
+    double result = sum(axisArray, length);
+    return result / length;    
+}
```
Comment (humleflue) [inline c-svc23Vectors/math.c:104]: change double to int

File: c-svc23Vectors/math.c [hunk 11]
```diff
@@ -34,3 +36,94 @@ int dotProductVectors(const int *input1, const int *input2, int length) {
...
+    double result = 0;
+    int i;
+    for (i = 0; i < length; i++)
+    {
+        if(axisArray[i] < 0)
+        {
+            result += -1 * axisArray[i];
+        } else {
+            result += axisArray[i];
+        }
```
Comment (humleflue) [inline c-svc23Vectors/math.c:116]:
```suggestion
        result += abs(axisArray[i]);
```

File: c-svc23Vectors/math.c [hunk 12]
```diff
@@ -34,3 +36,94 @@ int dotProductVectors(const int *input1, const int *input2, int length) {
...
+}
+
+double mean(double axisArray[], int length){
+    double result = sum(axisArray, length);
+    return result / length;    
+}
+
+double absSum(double axisArray[], int length){
+    double result = 0;
+    int i;
```
Comment (humleflue) [inline c-svc23Vectors/math.c:108]:
```suggestion
    int i, cur;
```

File: c-svc23Vectors/math.c [hunk 13]
```diff
@@ -34,3 +36,94 @@ int dotProductVectors(const int *input1, const int *input2, int length) {
...
+    return result / length;    
+}
+
+double mean(double axisArray[], int length){
+    double result = sum(axisArray, length);
+    return result / length;    
+}
+
+double absSum(double axisArray[], int length){
+    double result = 0;
```
Comment (humleflue) [inline c-svc23Vectors/math.c:107]: Change double to int

File: c-svc23Vectors/math.c [hunk 14]
```diff
@@ -34,3 +36,94 @@ int dotProductVectors(const int *input1, const int *input2, int length) {
...
+            result += -1 * axisArray[i];
+        } else {
+            result += axisArray[i];
+        }
+    }
+    return result;   
+}
+
+double sum(double axisArray[], int length){
+    double result = 0;
```
Comment (humleflue) [inline c-svc23Vectors/math.c:122]: double -> int

File: c-svc23Vectors/test/MathTests.c [hunk 15]
```diff
@@ -42,11 +42,216 @@ void dotProduct3dVector_givenVectorsOfIntegers_expectsTheCorrectDotProduct(CuTes
 	}
 */
 
+void getBiggestNumber_givenArrayOfNegativeNumbers_ShouldReturnNumberWithLowestNegativeValue(CuTest* tc)
+{
+	/* Arrange */
+    double numbers[TIME_SERIES_VECTORS_AMOUNT] = {-5,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3};
+
+	/* Act */
+	double expected = -5;
```
Comment (humleflue) [inline c-svc23Vectors/test/MathTests.c:51]: This belongs in the arrange step

File: c-svc23Vectors/test/MathTests.c [hunk 16]
```diff
@@ -42,11 +42,216 @@ void dotProduct3dVector_givenVectorsOfIntegers_expectsTheCorrectDotProduct(CuTes
...
+	CuAssertTrue(tc, expected == actual);
+}
+
+void getBiggestNumber_givenArrayOfPositiveNumbers_ShouldReturnNumberWithHighestPositiveValue(CuTest* tc)
+{
+	/* Arrange */
+    double numbers[TIME_SERIES_VECTORS_AMOUNT] = {5,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3};
+
+	/* Act */
+	double expected = 5;
```
Comment (humleflue) [inline c-svc23Vectors/test/MathTests.c:65]: This belongs in the arrange step

File: c-svc23Vectors/test/MathTests.c [hunk 17]
```diff
@@ -42,11 +42,216 @@ void dotProduct3dVector_givenVectorsOfIntegers_expectsTheCorrectDotProduct(CuTes
...
+    double actual = getBiggestNumber(numbers, TIME_SERIES_VECTORS_AMOUNT);
+
+
+	/* Assert */
+	CuAssertTrue(tc, expected == actual);
```
Comment (humleflue) [inline c-svc23Vectors/test/MathTests.c:70]: Use CuAssertIntEquals

File: c-svc23Vectors/test/MathTests.c [hunk 18]
```diff
@@ -42,11 +42,216 @@ void dotProduct3dVector_givenVectorsOfIntegers_expectsTheCorrectDotProduct(CuTes
...
+}
+
+void getBiggestNumber_givenArrayOfNegativeAndPositiveNumbers_ShouldReturnNumberWithBiggestAbosluteValue(CuTest* tc)
+{
+	/* Arrange */
+    double numbers[TIME_SERIES_VECTORS_AMOUNT] = {4,2,3,1,2,3,1,2,3,1,2,3,1,2,3,-5,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3};
+
+	/* Act */
+	double expected = -5;
```
Comment (humleflue) [inline c-svc23Vectors/test/MathTests.c:79]: this is arrange

File: c-svc23Vectors/test/MathTests.c [hunk 19]
```diff
@@ -42,11 +42,216 @@ void dotProduct3dVector_givenVectorsOfIntegers_expectsTheCorrectDotProduct(CuTes
...
+    double actual = getBiggestNumber(numbers, TIME_SERIES_VECTORS_AMOUNT);
+
+
+	/* Assert */
+	CuAssertTrue(tc, expected == actual);
```
Comment (humleflue) [inline c-svc23Vectors/test/MathTests.c:84]: use CuAssertIntEquals

File: c-svc23Vectors/test/MathTests.c [hunk 20]
```diff
@@ -42,11 +42,216 @@ void dotProduct3dVector_givenVectorsOfIntegers_expectsTheCorrectDotProduct(CuTes
...
+}
+
+void getSmallestNumber_givenArrayOfNegativeNumbers_ShouldReturnNumberWithHighestNegativeValue(CuTest* tc)
+{
+	/* Arrange */
+    double numbers[TIME_SERIES_VECTORS_AMOUNT] = {-5,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3};
+
+	/* Act */
+	double expected = -1;
```
Comment (humleflue) [inline c-svc23Vectors/test/MathTests.c:93]: arrange

File: c-svc23Vectors/test/MathTests.c [hunk 21]
```diff
@@ -42,11 +42,216 @@ void dotProduct3dVector_givenVectorsOfIntegers_expectsTheCorrectDotProduct(CuTes
...
+    double actual = getSmallestNumber(numbers, TIME_SERIES_VECTORS_AMOUNT);
+
+
+	/* Assert */
+	CuAssertTrue(tc, expected == actual);
```
Comment (humleflue) [inline c-svc23Vectors/test/MathTests.c:98]: CuAssertIntEquals

File: c-svc23Vectors/test/MathTests.c [hunk 22]
```diff
@@ -42,11 +42,216 @@ void dotProduct3dVector_givenVectorsOfIntegers_expectsTheCorrectDotProduct(CuTes
...
+}
+
+void getSmallestNumber_givenArrayOfPositiveNumbers_ShouldReturnNumberWithLowestPositiveValue(CuTest* tc)
+{
+	/* Arrange */
+    double numbers[TIME_SERIES_VECTORS_AMOUNT] = {5,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3,1,2,3};
+
+	/* Act */
+	double expected = 1;
```
Comment (humleflue) [inline c-svc23Vectors/test/MathTests.c:107]: arrange

File: c-svc23Vectors/test/MathTests.c [hunk 23]
```diff
@@ -42,11 +42,216 @@ void dotProduct3dVector_givenVectorsOfIntegers_expectsTheCorrectDotProduct(CuTes
...
+    double actual = getSmallestNumber(numbers, TIME_SERIES_VECTORS_AMOUNT);
+
+
+	/* Assert */
+	CuAssertTrue(tc, expected == actual);
```
Comment (humleflue) [inline c-svc23Vectors/test/MathTests.c:112]: CuAssertIntEquals

File: c-svc23Vectors/test/MathTests.c [hunk 24]
```diff
@@ -42,11 +42,216 @@ void dotProduct3dVector_givenVectorsOfIntegers_expectsTheCorrectDotProduct(CuTes
...
+}
+
+void getSmallestNumber_givenArrayOfNegativeAndPositiveNumbers_ShouldReturnNumberWithSmallestAbosluteValue(CuTest* tc)
+{
+	/* Arrange */
+    double numbers[TIME_SERIES_VECTORS_AMOUNT] = {4,2,3,1,2,3,1,2,3,1,2,3,1,2,3,-5,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3};
+
+	/* Act */
+	double expected = 1;
```
Comment (humleflue) [inline c-svc23Vectors/test/MathTests.c:121]: arrange

File: c-svc23Vectors/test/MathTests.c [hunk 25]
```diff
@@ -42,11 +42,216 @@ void dotProduct3dVector_givenVectorsOfIntegers_expectsTheCorrectDotProduct(CuTes
...
+    double actual = getSmallestNumber(numbers, TIME_SERIES_VECTORS_AMOUNT);
+
+
+	/* Assert */
+	CuAssertTrue(tc, expected == actual);
```
Comment (humleflue) [inline c-svc23Vectors/test/MathTests.c:126]: CuAssertIntEquals

File: c-svc23Vectors/test/MathTests.c [hunk 26]
```diff
@@ -42,11 +42,216 @@ void dotProduct3dVector_givenVectorsOfIntegers_expectsTheCorrectDotProduct(CuTes
...
+	/* Arrange */
+    double numbers[3] = {1.5, 1.5, 3};
+
+	/* Act */
+	double expected = 2;
+    double actual = absMean(numbers, 3);
+
+
+	/* Assert */
+	CuAssertTrue(tc, expected == actual);
```
Comment (humleflue) [inline c-svc23Vectors/test/MathTests.c:140]: CuAssertIntEquals

File: c-svc23Vectors/test/MathTests.c [hunk 27]
```diff
@@ -42,11 +42,216 @@ void dotProduct3dVector_givenVectorsOfIntegers_expectsTheCorrectDotProduct(CuTes
...
+}
+
+void absMean_givenAllPositives_shouldReturnCorrectMean(CuTest* tc)
+{
+	/* Arrange */
+    double numbers[3] = {1.5, 1.5, 3};
+
+	/* Act */
+	double expected = 2;
```
Comment (humleflue) [inline c-svc23Vectors/test/MathTests.c:135]: arrange

File: c-svc23Vectors/test/MathTests.c [hunk 28]
```diff
@@ -42,11 +42,216 @@ void dotProduct3dVector_givenVectorsOfIntegers_expectsTheCorrectDotProduct(CuTes
...
+	CuAssertTrue(tc, expected == actual);
+}
+
+void absMean_givenPositiveAndNegativeNumbers_shouldReturnCorrectMean(CuTest* tc)
+{
+	/* Arrange */
+    double numbers[3] = {-1.5, 1.5, -3};
+
+	/* Act */
+	double expected = 2;
```
Comment (humleflue) [inline c-svc23Vectors/test/MathTests.c:149]: arrange
- you get the point, right? 😅

File: c-svc23Vectors/test/MathTests.c [hunk 29]
```diff
@@ -42,11 +42,216 @@ void dotProduct3dVector_givenVectorsOfIntegers_expectsTheCorrectDotProduct(CuTes
...
+    double actual = absMean(numbers, 3);
+
+
+	/* Assert */
+	CuAssertTrue(tc, expected == actual);
```
Comment (humleflue) [inline c-svc23Vectors/test/MathTests.c:154]: CuAssertIntEquals

File: c-svc23Vectors/test/TransformTests.c [hunk 30]
```diff
@@ -85,6 +85,29 @@ void transform_transformTso_xYZShouldBeAverage(CuTest* tc)
...
+void extractXYZ_transformTso_xYZShouldBeSplitInArrays(CuTest* tc){
+	/* Arrange */
+	TimeSeriesObservation tso = {true, TIME_SERIES_OBSERVATION_INIT_108_VALUES, 0};
+	int i;
+	double x[TIME_SERIES_VECTORS_AMOUNT];
+	double y[TIME_SERIES_VECTORS_AMOUNT];
+	double z[TIME_SERIES_VECTORS_AMOUNT];
+	double expectedX[TIME_SERIES_VECTORS_AMOUNT] = {1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1};
+	double expectedY[TIME_SERIES_VECTORS_AMOUNT] = {2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2};
+	double expectedZ[TIME_SERIES_VECTORS_AMOUNT] = {3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3};
```
Comment (humleflue) [inline c-svc23Vectors/test/TransformTests.c:97]: int

File: c-svc23Vectors/test/MathTests.c [hunk 31]
```diff
@@ -42,11 +42,216 @@ void dotProduct3dVector_givenVectorsOfIntegers_expectsTheCorrectDotProduct(CuTes
 	}
 */
 
+void getBiggestNumber_givenArrayOfNegativeNumbers_ShouldReturnNumberWithLowestNegativeValue(CuTest* tc)
+{
+	/* Arrange */
+    double numbers[TIME_SERIES_VECTORS_AMOUNT] = {-5,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3,-1,-2,-3};
```
Comment (humleflue) [inline c-svc23Vectors/test/MathTests.c:48]: int

File: c-svc23Vectors/test/MathTests.c (hunk 15 above)
Comment (humleflue) [reply c-svc23Vectors/test/MathTests.c:51]: and its an int

File: c-svc23Vectors/test/TransformTests.c [hunk 32]
```diff
@@ -85,6 +85,29 @@ void transform_transformTso_xYZShouldBeAverage(CuTest* tc)
...
+
+	/* Act */
+	extractXYZ(tso, x, y, z);
+
+	/* Assert */
+	for (i = 0; i < TIME_SERIES_VECTORS_AMOUNT; i++)
+	{
+		CuAssertTrue(tc, expectedX[i] == x[i]);
+		CuAssertTrue(tc, expectedY[i] == y[i]);
+		CuAssertTrue(tc, expectedZ[i] == z[i]);
```
Comment (humleflue) [inline c-svc23Vectors/test/TransformTests.c:107]: CuAssertIntEquals

File: c-svc23Vectors/types.h [hunk 33]
```diff
@@ -4,12 +4,12 @@
...
 #define ACCELEROMETOR_VECTOR_DIM 3
-#define TIME_SERIES_VECTORS_AMOUNT 12
+#define TIME_SERIES_VECTORS_AMOUNT 36
 #define TIME_SERIES_OBSERVATION_DIM (TIME_SERIES_VECTORS_AMOUNT * ACCELEROMETOR_VECTOR_DIM)
-#define SAMPLED_OBSERVATION_DIM 3
+#define SAMPLED_OBSERVATION_DIM 23
 /* The most important thing here is that _currentIndex is set to 0 */
 #define TIME_SERIES_OBSERVATION_INIT { false, { 0 }, 0 }
-
+#define TIME_SERIES_OBSERVATION_INIT_108_VALUES {1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3}
```
Comment (humleflue) [inline c-svc23Vectors/types.h:12]: If this is only used in tests (which I assume), then please put it in the test file.

**Review by humleflue — CHANGES_REQUESTED** [review]
Looks good. I have put some comments for polishing. I love the tests! 😍

**Comment by humleflue** [comment]
If you pull main, I think your pipeline will pass.

File: c-svc23Vectors/math.c (hunk 6 above)
Comment (humleflue) [reply c-svc23Vectors/math.c:63]: Hvis det ikke er det her, som du vil, men at du gerne vil beholde dit sign, så synes jeg at det er super det her, men så må du gerne lige formatere det ordenligt 😉

File: c-svc23Vectors/math.c [hunk 34]
```diff
@@ -1,4 +1,8 @@
 #include "math.h"
+#include "types.h"
+
+/* This is used as an array for intermediate results */
+double differencesSquared[TIME_SERIES_OBSERVATION_DIM];
```
Comment (humleflue) [inline c-svc23Vectors/math.c:5]: Move this into absVariance and put a comment, that this was not transfered as a parameter as the Kin would not allow us to do so.

File: c-svc23Vectors/math.c [hunk 35]
```diff
@@ -35,6 +37,162 @@ int dotProduct3dVectors(const int input1[3], const int input2[3]) {
...
+
+    xyz[0] = smallestX;
+    xyz[1] = smallestY;
+    xyz[2] = smallestZ;
+        
+    return getSmallestAbsNumberAsSigned(xyz, 3);
+}
+
+int getSmallestAbsNumberAsSigned(int axisArray[], int length){
+    int smallestNumber = 999999;
```
Comment (humleflue) [inline c-svc23Vectors/math.c:93]: Lav symbolsk konstant INT_MAX (behøver ikke at være INT_MAX men bare et stort tal)

**Review by humleflue — APPROVED** [review]

File: c-svc23Vectors/math.c (hunk 35 above)
Comment (laurits-mumberg) [reply c-svc23Vectors/math.c]: DONE

**Pull request merged**